### PR TITLE
When starting a workspace but usage attribution is unclear, prompt for explicit user choice

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

When starting a workspace but usage attribution is unclear, prompt for explicit user choice.

Possible unclear situations:
- User is a member of multiple teams with usage-based billing enabled
- User explicitly chose to attribute to a team, but they've since left this team (or the team was deleted, or its subscription was cancelled)

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/11579

## How to test

1. Start a workspace on any repo (should work)
2. Create a team called "Gitpod" (or "Gitpod 1", "Gitpod 2"... whichever is available) to enable the Usage-Based Billing flag
3. Start a new workspace on any repo (should work)
4. Upgrade your team to Usage-Based Billing by adding a payment method (e.g. 4242 4242 4242 4242, 4/24, 424)
5. Start a new workspace on any repo (should still work)
6. Create a new team (called anything you want)
7. Start a new workspace (should work)
8. Upgrade your second team to usage-based billing
9. Start a new workspace -- this time it should fail (unclear which team to bill) and ask you for an explicit choice
10. Select your second team -- this should retry the workspace start, and succeed
11. Delete your second team
12. Start a new workspace -- this time it should fail again (explicit choice is no longer valid) and ask you for a new choice

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview
- [x] /werft with-payment

### CODE REVIEW

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 1]
```diff
@@ -2008,20 +2008,6 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
...
-            const members = await this.teamDB.findMembersByTeam(teamId);
-            await Promise.all(
-                members.map(async (m) => {
-                    const u = await this.userDB.findUserById(m.userId);
-                    if (u && !u.usageAttributionId) {
-                        u.usageAttributionId = attributionId;
-                        await this.userDB.storeUser(u);
-                    }
-                }),
-            );
```
Comment (andrew-farries) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:2024]: Why is this removed?

File: components/server/src/user/user-service.ts [hunk 2]
```diff
@@ -229,12 +229,16 @@ export class UserService {
...
+                // TODO(janx): Throw error if selected attribution ID is no longer valid (e.g. user left team, team deleted, team cancelled)
+                return user.usageAttributionId;
             }
-            // Return the user's explicit attribution ID.
-            return user.usageAttributionId;
+            // TODO(janx):
+            //   - If single team with Usage-Based billing enabled, auto-attribute to that team
+            //   - If multiple teams with Usage-Based, throw error to prompt for explicit user choice
+            // No explicit user attribution ID yet -- attribute all usage to the user by default (regardless of project/team).
+            return AttributionId.render({ kind: "user", userId: user.id });
```
Comment (andrew-farries) [inline components/server/src/user/user-service.ts:241]: There is no change here right? Just inverting the `if/else` and adding TODOs.

**Review by andrew-farries — COMMENTED** [review]
PR description doesn't seem to match the code 😄 

AIUI this PR just removes the automatic attribution when a team signs up for Stripe.

**Comment by jankeromnes** [comment]
Whoops, this was meant to be a Draft. 😅 Thanks for the quick review though @andrew-farries! 🙏

File: components/server/ee/src/workspace/gitpod-server-impl.ts (hunk 1 above)
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:2024]: The logic of this change is:
1. Stop setting/unsetting `User.usageAttributionId` automatically, but give full explicit control of this value to the user
2. When attributing instances on start, we already look at the explicit user attribution, but if none is set and attribution, or it's set to an invalid choice, we should either guess what the sensible non-ambiguous attribution should be, or throw an error in the backend and show a prompt in the front-end in order to ask the user what their explicit choice is

Step 1. was partly implemented by https://github.com/gitpod-io/gitpod/pull/11584, but this particular location was forgotten so I'm removing it now

Step 2. is to be implemented in this PR (the whole non-ambiguous logic, error case, and Modal shown to the user)

File: components/server/src/user/user-service.ts [hunk 3]
```diff
@@ -229,12 +292,18 @@ export class UserService {
     async getWorkspaceUsageAttributionId(user: User, projectId?: string): Promise<string | undefined> {
         // A. Billing-based attribution
```
Comment (jankeromnes) [inline components/server/src/user/user-service.ts:293]: TODO: Handle UBB flag off (fall back to Project-based attribution? or always to user?)

**Comment by jankeromnes** [comment]
@gtsiolis If you have some time, could you please provide some advice on the UX? 🙂

I went with a very basic, non-dismissible Modal, but it's still rough around the edges:

| Starting a Workspace | User Billing |
| --- | --- |
| Screenshot 2022-08-03 at 15 05 47 | Screenshot 2022-08-03 at 15 06 43 |
| Screenshot 2022-08-03 at 15 06 09 | Screenshot 2022-08-03 at 15 06 26 |

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/components/BillingAccountSelector.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import { getGitpodService } from "../service/service";
+import { TeamsContext } from "../teams/teams-context";
+import { UserContext } from "../user-context";
+import SelectableCardSolid from "../components/SelectableCardSolid";
+import { ReactComponent as Spinner } from "../icons/Spinner.svg";
+
+export function BillingAccountSelector(props: { onSelected?: () => void }) {
+    const { user, setUser } = useContext(UserContext);
+    const { teams } = useContext(TeamsContext);
+    const [teamsWithBillingEnabled, setTeamsWithBillingEnabled] = useState<Team[] | undefined>();
```
Comment (AlexTugarev) [inline components/dashboard/src/components/BillingAccountSelector.tsx:18]: @jankeromnes, I bet this is not the single place, where we'd need this kind of info. have you considered promoting this flag to the result of getTeams? `team.isBillingEnabled` would be nice and it shave off the additional requests.

File: components/dashboard/src/components/BillingAccountSelector.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                }
+            }),
+        ).then(() => setTeamsWithBillingEnabled(teamsWithBilling));
+    }, [teams]);
+
+    const setUsageAttributionTeam = async (team?: Team) => {
+        if (!user) {
+            return;
+        }
+        const usageAttributionId = team ? `team:${team.id}` : `user:${user.id}`;
```
Comment (AlexTugarev) [inline components/dashboard/src/components/BillingAccountSelector.tsx:40]: hint: there is a `AttributionId.render`

File: components/dashboard/src/start/CreateWorkspace.tsx [hunk 6]
```diff
@@ -185,6 +186,12 @@ export default class CreateWorkspace extends React.Component<CreateWorkspaceProp
                     phase = StartPhase.Stopped;
                     statusMessage = <LimitReachedOutOfHours />;
                     break;
+                case ErrorCodes.INVALID_COST_CENTER:
+                    // HACK: Hide the error (behind the modal)
+                    error = undefined;
+                    phase = StartPhase.Stopped;
+                    statusMessage = <SelectCostCenterModal onSelected={() => this.createWorkspace()} />;
```
Comment (AlexTugarev) [inline components/dashboard/src/start/CreateWorkspace.tsx:193]: `onSelected => createWorkspace()` doesn't seem to invalidate error state, I think this might delay the visual effect, e.g. the transition from the modal to next state. maybe we should clear the error state therein, or try to reload when selection was done.

File: components/gitpod-protocol/src/messaging/error.ts [hunk 7]
```diff
@@ -32,15 +32,18 @@ export namespace ErrorCodes {
     // 430 Repository not whitelisted (custom status code)
     export const REPOSITORY_NOT_WHITELISTED = 430;
 
+    // 450 Payment error
+    export const PAYMENT_ERROR = 450;
+
+    // 455 Invalid cost center (custom status code)
+    export const INVALID_COST_CENTER = 455;
```
Comment (AlexTugarev) [inline components/gitpod-protocol/src/messaging/error.ts:39]: can we agree on a common prefix for everything payment related, please?

also see https://github.com/gitpod-io/gitpod/pull/11576/files#diff-3bb6d890c14ac42bf9860578a9f6a85ffd0121dbf980055baa99f8b7731c46daR45

and feel free to comment on that BTW!

alternative way to approach this, we could use just `PAYMENT_ERROR` (in similar cases) and encode specific conditions into the payload. it feels like we'll have the needs to extends for more error modes. wdyt?

File: components/dashboard/src/components/BillingAccountSelector.tsx (hunk 4 above)
Comment (jankeromnes) [reply components/dashboard/src/components/BillingAccountSelector.tsx:19]: @AlexTugarev I think this would be premature, because:
- I can't imagine other places that will need the same information currently (note: this billing account selector component is already reusable, and in fact it's already used in two places)
- I don't think it's wise to make `getTeams` load all the team subscriptions on every call, even when most callers don't need that currently

File: components/gitpod-protocol/src/messaging/error.ts (hunk 7 above)
Comment (jankeromnes) [reply components/gitpod-protocol/src/messaging/error.ts:39]: I don't see much value in aligning every payment-related error around a single prefix or error code range (apart from readability of the error codes list -- which we arguably don't read that often).

Personally, I'd prefer to leave this as is (unless you have a strong opinion towards a different error name or code), and you can use whichever convention seems best in your PR. We can always align later if you think it's valuable, but I suspect that aligning these is not important in the grand scheme of things. 😇

**Comment by jankeromnes** [comment]
Thanks for the reviews!

I guess this is now ready to be merged, unless you have any final requests. 🙂

File: components/server/src/user/user-service.ts [hunk 8]
```diff
@@ -213,6 +216,66 @@ export class UserService {
...
+            if (!subscriptionId) {
+                throw new ResponseError(
+                    ErrorCodes.INVALID_COST_CENTER,
+                    "The billing team you've selected has no active subscription.",
+                );
+            }
+        }
+    }
+
+    protected async findSingleTeamWithUsageBasedBilling(user: User): Promise<Team | undefined> {
```
Comment (geropl) [inline components/server/src/user/user-service.ts:255]: FYI: This can also come out of `BillingMode`: we already gather the data there, and only need to loop it through.

But that's improvement after both got merged.

File: components/server/src/user/user-service.ts [hunk 9]
```diff
@@ -231,12 +294,18 @@ export class UserService {
     async getWorkspaceUsageAttributionId(user: User, projectId?: string): Promise<string | undefined> {
```
Comment (geropl) [inline components/server/src/user/user-service.ts:294]: nit, and out-of-scope here: It feels like this method (and it's descendants) do not necessarily belong into UserService, but more into `usage/Attribution` or so. But that's sth for later. :upside_down_face:

**Comment by geropl** [comment]
@jankeromnes testing this now...

**Comment by geropl** [comment]
On re-start, the chooser is not shown, but instead I see the plain error:
image

I think that's what we discussed earlier, and part of a follow-up PR.

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works :skateboard: 

[Error message](https://github.com/gitpod-io/gitpod/pull/11777#issuecomment-1206304264) is not nice, but does not harm prod, and can be a follow-up. :100:
